### ISSUE TITLE
Infinite recursion error when using $ to consume entire input type

### ISSUE THREAD

**Issue by jakobo** [typed-graphql-builder/typed-graphql-builder#29]
A change in 2.2.0 broke type inference for non-primitive types, even when using the new `$$` syntax to force a required input. I did some basic digging into the generated file, and it seems to be linked to the `VariabledInput` changes.

**Proposed Fix**: Add `Variable<INPUT_TYPE,any>` to the generated input types, allowing an entire input type to be consumed as a Variable.

Reproduction:

1. Hasura style graphql schema (any works)
2. Construct a mutation that takes an entire input type as a variable

```ts
const upsertJobMutation = mutation((m) => [
  m.insert_job_one(
    {
      object: $$("job")
    },
    (j) => [j.id]
  ),
]);
```

Generated code:
```ts
/**
 * insert a single row into the table: "jobs"
 */
insert_job_one<Args extends VariabledInput<{
  object: job_insert_input
  on_conflict?: job_on_conflict | null | undefined,
}>,Sel extends Selection<job>>(args: Args, selectorFn: (s: job) => [...Sel]):$Field<"insert_job_one", GetOutput<Sel> | undefined , GetVariables<Sel, Args>> {

  const options = {
    argTypes: {
          object: "job_insert_input!",
          on_conflict: "job_on_conflict"
        },
    args,

    selection: selectorFn(new job)
  };
  return this.$_select("insert_job_one", options) as any
}
```

Working Code:
```ts
/**
 * insert a single row into the table: "jobs"
 */
insert_job_one<Args extends VariabledInput<{
  object: job_insert_input | Variable<job_insert_input, any>
  on_conflict?: job_on_conflict | null | undefined  | Variable<job_on_conflict, any>,
}>,Sel extends Selection<job>>(args: Args, selectorFn: (s: job) => [...Sel]):$Field<"insert_job_one", GetOutput<Sel> | undefined , GetVariables<Sel, Args>> {

  const options = {
    argTypes: {
          object: "job_insert_input!",
          on_conflict: "job_on_conflict"
        },
    args,

    selection: selectorFn(new job)
  };
  return this.$_select("insert_job_one", options) as any
}
```

Expected: Can consume `$$("job")` as a matching input of `job_insert_input`

Actual: TS error `Type instantiation is excessively deep and possibly infinite.ts(2589)`

**Comment by jakobo** [comment]
As a followup, I _think_ it can be addressed in the compile step when we print input fields, but I'm still working to familiarize myself with the code base.

https://github.com/typed-graphql-builder/typed-graphql-builder/blob/fd8dcce9719b90bd88720905a21b02d0fdac2b9b/src/compile.ts#L172

```ts
return `${def.name.value}${maybe}: ${printType(def.type)} | Variable<${printType(def.type)}>`
```

**Comment by spion** [comment]
I found the reason for the issue. A trick I use to debug these is to either remove the recursive input type to see the issue, or to switch to a limited-depth definition (but that one doesn't support variables past that depth)

```typescript
type Decrement<X extends number> = X extends 5
  ? 4
  : X extends 4
  ? 3
  : X extends 3
  ? 2
  : X extends 2
  ? 1
  : 0

type VariabledInput<T, MaxDepth extends number = 5> = MaxDepth extends 0
  ? T
  : [T] extends [$Atomic | null | undefined]
  ? Variable<T, any> | T
  : T extends ReadonlyArray<infer I>
  ? Variable<T, any> | T | ArrayInput<I>
  : T extends Record<string, any>
  ? { [K in keyof T]: VariabledInput<T[K], Decrement<MaxDepth>> } | T
  : never
```

It turned out that when `T extends Record<string, any>`, we don't allow Variable<T, any> as the option anymore. Should be a one-line fix.

I wonder if it makes sense to limit the recursion depth to get better error messages? When I limit it to 5, I get the following error:

```
examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ [K in keyof booking_channel_bool_exp]: ExtractInputVariables<booking_channel_bool_exp[K]>; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | ...)[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: Strin...'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | { ...; })[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: ...'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ _and?: any; _not?: any; _or?: any; name?: unknown; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_not' circularly references itself in mapped type '{ [K in keyof booking_channel_bool_exp]: ExtractInputVariables<booking_channel_bool_exp[K]>; }'.
examples/test-x.good.ts(46,3): error TS2615: Type of property '_not' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | ...)[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: Strin...'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_not' circularly references itself in mapped type '{ _and?: any; _not?: any; _or?: any; name?: unknown; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_or' circularly references itself in mapped type '{ [K in keyof booking_channel_bool_exp]: ExtractInputVariables<booking_channel_bool_exp[K]>; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_or' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | ...)[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: Strin...'.

examples/test-x.good.ts(48,7): error TS2322: Type 'Variable<{}, "job">' is not assignable to type 'booking_channel_insert_input | { name?: string | Variable<string | null | undefined, any> | null | undefined; }'.
```

Notice the last message - when the recursion depth is limited, we also get the actual error at the end.

**Comment by spion** [comment]
But yeah, I see your point. For the best experience, it may be a good idea to "pre-apply" VariabledInput as a preprocessing function for all InputTypes. I'll give it some thought - meanwhile I'll do a one-liner patch to fix support for object variables.

**Comment by spion** [comment]
Fix published in v2.2.3

**Comment by jakobo** [comment]
Can confirm that resolved it. Will let you know if I bump into anything else. For future reference, our stack at work is:

- urql (both via react, or @urql/core for backend services), calling
- tgqlb for typed graphql queries (formerly graphql-code-generator), talking to
- hasura (non-relay interface)

Also, appreciate the info RE type inference! I've never seen the decrement pattern... that's seriously amazing. 🤯
